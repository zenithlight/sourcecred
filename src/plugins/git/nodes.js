// @flow

import deepFreeze from "deep-freeze";
import {NodeAddress, type NodeAddressT} from "../../core/graph";
import type {Hash} from "./types";

export opaque type RawAddress: NodeAddressT = NodeAddressT;

const GIT_PREFIX = NodeAddress.fromParts(["sourcecred", "git"]);
export function _gitAddress(...parts: string[]): RawAddress {
  return NodeAddress.append(GIT_PREFIX, ...parts);
}

export const COMMIT_TYPE: "COMMIT" = "COMMIT";

export const Prefix: {|base: NodeAddressT, commit: RawAddress|} = deepFreeze({
  base: GIT_PREFIX,
  commit: _gitAddress(COMMIT_TYPE),
});

export type CommitAddress = {|
  +type: typeof COMMIT_TYPE,
  +hash: Hash,
|};

export type StructuredAddress = CommitAddress;

export function fromRaw(x: RawAddress): StructuredAddress {
  function fail() {
    return new Error(`Bad address: ${NodeAddress.toString(x)}`);
  }
  if (!NodeAddress.hasPrefix(x, GIT_PREFIX)) {
    throw fail();
  }
  const [_unused_sc, _unused_git, _type, ...rest] = NodeAddress.toParts(x);
  const type: $ElementType<StructuredAddress, "type"> = (_type: any);
  switch (type) {
    case "COMMIT": {
      if (rest.length !== 1) throw fail();
      const [hash] = rest;
      return {type: COMMIT_TYPE, hash};
    }
    default:
      // eslint-disable-next-line no-unused-expressions
      (type: empty);
      throw fail();
  }
}

export function toRaw(x: StructuredAddress): RawAddress {
  switch (x.type) {
    case COMMIT_TYPE:
      return NodeAddress.append(Prefix.commit, x.hash);
    default:
      throw new Error(`Unexpected type ${(x.type: empty)}`);
  }
}
