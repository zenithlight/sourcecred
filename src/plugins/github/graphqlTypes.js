// @flow

// Autogenerated file. Do not edit.

import deepFreeze from "deep-freeze";

export type Actor = Bot | Organization | User;

export type Blob = {|
  +__typename: "Blob",
  +id: string,
  +oid: GitObjectID,
|};

export type Bot = {|
  +__typename: "Bot",
  +id: string,
  +login: String,
  +url: URI,
|};

export type Commit = {|
  +__typename: "Commit",
  +author: null | {|
    +date: null | GitTimestamp,
    +user: null | Bot | User,
  |},
  +authoredDate: GitTimestamp,
  +id: string,
  +message: String,
  +oid: GitObjectID,
  +parents: $ReadOnlyArray<null | Commit>,
  +url: URI,
|};

export type DateTime = string;

export type GitObject = Blob | Commit | Tag | Tree;

export type GitObjectID = string;

export type GitTimestamp = string;

export type Int = number;

export type Issue = {|
  +__typename: "Issue",
  +author: null | Actor,
  +body: String,
  +comments: $ReadOnlyArray<null | IssueComment>,
  +createdAt: DateTime,
  +id: string,
  +number: Int,
  +reactions: $ReadOnlyArray<null | Reaction>,
  +title: String,
  +url: URI,
|};

export type IssueComment = {|
  +__typename: "IssueComment",
  +author: null | Actor,
  +body: String,
  +createdAt: DateTime,
  +id: string,
  +reactions: $ReadOnlyArray<null | Reaction>,
  +url: URI,
|};

export type Organization = {|
  +__typename: "Organization",
  +id: string,
  +login: String,
  +url: URI,
|};

export type PullRequest = {|
  +__typename: "PullRequest",
  +additions: Int,
  +author: null | Actor,
  +baseRefName: String,
  +body: String,
  +comments: $ReadOnlyArray<null | IssueComment>,
  +createdAt: DateTime,
  +deletions: Int,
  +id: string,
  +mergeCommit: null | Commit,
  +number: Int,
  +reactions: $ReadOnlyArray<null | Reaction>,
  +reviews: $ReadOnlyArray<null | PullRequestReview>,
  +title: String,
  +url: URI,
|};

export type PullRequestReview = {|
  +__typename: "PullRequestReview",
  +author: null | Actor,
  +body: String,
  +comments: $ReadOnlyArray<null | PullRequestReviewComment>,
  +createdAt: DateTime,
  +id: string,
  +state: PullRequestReviewState,
  +url: URI,
|};

export type PullRequestReviewComment = {|
  +__typename: "PullRequestReviewComment",
  +author: null | Actor,
  +body: String,
  +createdAt: DateTime,
  +id: string,
  +reactions: $ReadOnlyArray<null | Reaction>,
  +url: URI,
|};

export type PullRequestReviewState =
  | "APPROVED"
  | "CHANGES_REQUESTED"
  | "COMMENTED"
  | "DISMISSED"
  | "PENDING";

export const PullRequestReviewState$Values: {|
  +"APPROVED": "APPROVED",
  +"CHANGES_REQUESTED": "CHANGES_REQUESTED",
  +"COMMENTED": "COMMENTED",
  +"DISMISSED": "DISMISSED",
  +"PENDING": "PENDING",
|} = deepFreeze({
  "APPROVED": "APPROVED",
  "CHANGES_REQUESTED": "CHANGES_REQUESTED",
  "COMMENTED": "COMMENTED",
  "DISMISSED": "DISMISSED",
  "PENDING": "PENDING",
});

export type Reaction = {|
  +__typename: "Reaction",
  +content: ReactionContent,
  +createdAt: DateTime,
  +id: string,
  +user: null | Organization | User,
|};

export type ReactionContent =
  | "CONFUSED"
  | "EYES"
  | "HEART"
  | "HOORAY"
  | "LAUGH"
  | "ROCKET"
  | "THUMBS_DOWN"
  | "THUMBS_UP";

export const ReactionContent$Values: {|
  +"CONFUSED": "CONFUSED",
  +"EYES": "EYES",
  +"HEART": "HEART",
  +"HOORAY": "HOORAY",
  +"LAUGH": "LAUGH",
  +"ROCKET": "ROCKET",
  +"THUMBS_DOWN": "THUMBS_DOWN",
  +"THUMBS_UP": "THUMBS_UP",
|} = deepFreeze({
  "CONFUSED": "CONFUSED",
  "EYES": "EYES",
  "HEART": "HEART",
  "HOORAY": "HOORAY",
  "LAUGH": "LAUGH",
  "ROCKET": "ROCKET",
  "THUMBS_DOWN": "THUMBS_DOWN",
  "THUMBS_UP": "THUMBS_UP",
});

export type Ref = {|
  +__typename: "Ref",
  +id: string,
  +target: null | GitObject,
|};

export type Repository = {|
  +__typename: "Repository",
  +createdAt: DateTime,
  +defaultBranchRef: null | Ref,
  +id: string,
  +issues: $ReadOnlyArray<null | Issue>,
  +name: String,
  +owner: null | RepositoryOwner,
  +pullRequests: $ReadOnlyArray<null | PullRequest>,
  +url: URI,
|};

export type RepositoryOwner = Organization | User;

export type String = string;

export type Tag = {|
  +__typename: "Tag",
  +id: string,
  +oid: GitObjectID,
|};

export type Tree = {|
  +__typename: "Tree",
  +id: string,
  +oid: GitObjectID,
|};

export type URI = string;

export type User = {|
  +__typename: "User",
  +id: string,
  +login: String,
  +url: URI,
|};
