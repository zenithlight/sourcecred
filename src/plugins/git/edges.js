// @flow

import deepFreeze from "deep-freeze";
import {
  type Edge,
  type EdgeAddressT,
  EdgeAddress,
  NodeAddress,
} from "../../core/graph";
import * as GitNode from "./nodes";
import {type TimestampMs} from "../../util/timestamp";

export opaque type RawAddress: EdgeAddressT = EdgeAddressT;

export const HAS_TREE_TYPE: "HAS_TREE" = "HAS_TREE";
export const HAS_PARENT_TYPE: "HAS_PARENT" = "HAS_PARENT";
export const INCLUDES_TYPE: "INCLUDES" = "INCLUDES";
export const BECOMES_TYPE: "BECOMES" = "BECOMES";
export const HAS_CONTENTS_TYPE: "HAS_CONTENTS" = "HAS_CONTENTS";

const GIT_PREFIX = EdgeAddress.fromParts(["sourcecred", "git"]);
function gitEdgeAddress(...parts: string[]): RawAddress {
  return EdgeAddress.append(GIT_PREFIX, ...parts);
}

export const Prefix: {|base: EdgeAddressT, hasParent: RawAddress|} = deepFreeze(
  {
    base: GIT_PREFIX,
    hasParent: gitEdgeAddress(HAS_PARENT_TYPE),
  }
);

export type HasParentAddress = {|
  type: typeof HAS_PARENT_TYPE,
  child: GitNode.CommitAddress,
  parent: GitNode.CommitAddress,
|};

export type StructuredAddress = HasParentAddress;

export const createEdge: {|
  hasParent: (
    child: GitNode.CommitAddress,
    parent: GitNode.CommitAddress,
    timestampMs: TimestampMs
  ) => Edge,
|} = deepFreeze({
  hasParent: (
    child: GitNode.CommitAddress,
    parent: GitNode.CommitAddress,
    timestampMs: TimestampMs
  ): Edge => ({
    address: toRaw({type: HAS_PARENT_TYPE, child, parent}),
    src: GitNode.toRaw(child),
    dst: GitNode.toRaw(parent),
    timestampMs,
  }),
});

const NODE_PREFIX_LENGTH = NodeAddress.toParts(GitNode._gitAddress()).length;

function lengthEncode(x: GitNode.RawAddress): $ReadOnlyArray<string> {
  const baseParts = NodeAddress.toParts(x).slice(NODE_PREFIX_LENGTH);
  return [String(baseParts.length), ...baseParts];
}
function lengthDecode(
  x: $ReadOnlyArray<string>,
  fail: () => Error
): {|+parts: $ReadOnlyArray<string>, +rest: $ReadOnlyArray<string>|} {
  if (x.length === 0) {
    // Not length-encoded.
    throw fail();
  }
  const [lengthString, ...allParts] = x;
  const length = parseInt(lengthString, 10);
  if (isNaN(length)) {
    throw fail();
  }
  if (length > allParts.length) {
    // Not enough elements.
    throw fail();
  }
  return {parts: allParts.slice(0, length), rest: allParts.slice(length)};
}
function multiLengthDecode(x: $ReadOnlyArray<string>, fail: () => Error) {
  let remaining = x;
  const partses = [];
  while (remaining.length > 0) {
    const {parts, rest} = lengthDecode(remaining, fail);
    partses.push(parts);
    remaining = rest;
  }
  return partses;
}

export function fromRaw(x: RawAddress): StructuredAddress {
  function fail() {
    return new Error(`Bad address: ${EdgeAddress.toString(x)}`);
  }
  if (!EdgeAddress.hasPrefix(x, GIT_PREFIX)) {
    throw fail();
  }
  const [_unused_sc, _unused_git, _type, ...rest] = EdgeAddress.toParts(x);
  const type: $ElementType<StructuredAddress, "type"> = (_type: any);
  switch (type) {
    case "HAS_PARENT": {
      const parts = multiLengthDecode(rest, fail);
      if (parts.length !== 2) throw fail();
      const [childParts, parentParts] = parts;
      const child: GitNode.CommitAddress = (GitNode.fromRaw(
        GitNode._gitAddress(...childParts)
      ): any);
      const parent: GitNode.CommitAddress = (GitNode.fromRaw(
        GitNode._gitAddress(...parentParts)
      ): any);
      return {type: HAS_PARENT_TYPE, child, parent};
    }
    default:
      // eslint-disable-next-line no-unused-expressions
      (type: empty);
      throw fail();
  }
}

export function toRaw(x: StructuredAddress): RawAddress {
  switch (x.type) {
    case HAS_PARENT_TYPE:
      return EdgeAddress.append(
        Prefix.hasParent,
        ...lengthEncode(GitNode.toRaw(x.child)),
        ...lengthEncode(GitNode.toRaw(x.parent))
      );
    default:
      throw new Error((x.type: empty));
  }
}
