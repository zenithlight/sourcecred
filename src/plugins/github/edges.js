// @flow

import deepFreeze from "deep-freeze";
import {
  type Edge,
  type EdgeAddressT,
  EdgeAddress,
  NodeAddress,
} from "../../core/graph";
import * as GithubNode from "./nodes";
import * as GitNode from "../git/nodes";
import {
  type ReactionContent,
  ReactionContent$Values as Reactions,
} from "./graphqlTypes";
import {type TimestampMs} from "../../util/timestamp";

export opaque type RawAddress: EdgeAddressT = EdgeAddressT;

export const AUTHORS_TYPE = "AUTHORS";
export const MERGED_AS_TYPE = "MERGED_AS";
export const HAS_PARENT_TYPE = "HAS_PARENT";
export const REFERENCES_TYPE = "REFERENCES";
export const REACTS_TYPE = "REACTS";
// GitHub tracks its own notion of a commit, which has a particular
// database id, is scoped to a particular repository, and has a canonical url
// on GitHub. The CORRESPONDS_TO_COMMIT_TYPE edges connect the GitHub commits
// to the corresponding Git commit.
export const CORRESPONDS_TO_COMMIT_TYPE = "CORRESPONDS_TO_COMMIT_TYPE";

const GITHUB_PREFIX = EdgeAddress.fromParts(["sourcecred", "github"]);
function githubEdgeAddress(...parts: string[]): RawAddress {
  return EdgeAddress.append(GITHUB_PREFIX, ...parts);
}

export const Prefix: {|
  authors: RawAddress,
  base: EdgeAddressT,
  correspondsToCommit: RawAddress,
  hasParent: RawAddress,
  mergedAs: RawAddress,
  reacts: RawAddress,
  reactsHeart: RawAddress,
  reactsHooray: RawAddress,
  reactsRocket: RawAddress,
  reactsThumbsUp: RawAddress,
  references: RawAddress,
|} = deepFreeze({
  base: GITHUB_PREFIX,
  authors: githubEdgeAddress(AUTHORS_TYPE),
  mergedAs: githubEdgeAddress(MERGED_AS_TYPE),
  references: githubEdgeAddress(REFERENCES_TYPE),
  hasParent: githubEdgeAddress(HAS_PARENT_TYPE),
  reacts: githubEdgeAddress(REACTS_TYPE),
  reactsThumbsUp: githubEdgeAddress(REACTS_TYPE, Reactions.THUMBS_UP),
  reactsHeart: githubEdgeAddress(REACTS_TYPE, Reactions.HEART),
  reactsHooray: githubEdgeAddress(REACTS_TYPE, Reactions.HOORAY),
  reactsRocket: githubEdgeAddress(REACTS_TYPE, Reactions.ROCKET),
  correspondsToCommit: githubEdgeAddress(CORRESPONDS_TO_COMMIT_TYPE),
});

export type AuthorsAddress = {|
  +type: typeof AUTHORS_TYPE,
  +author: GithubNode.UserlikeAddress,
  +content: GithubNode.AuthorableAddress,
|};
export type MergedAsAddress = {|
  +type: typeof MERGED_AS_TYPE,
  +pull: GithubNode.PullAddress,
|};
export type HasParentAddress = {|
  +type: typeof HAS_PARENT_TYPE,
  +child: GithubNode.ChildAddress,
|};
export type ReferencesAddress = {|
  +type: typeof REFERENCES_TYPE,
  +referrer: GithubNode.TextContentAddress,
  +referent: GithubNode.ReferentAddress,
|};
export type ReactsAddress = {|
  +type: typeof REACTS_TYPE,
  +reactionType: ReactionContent,
  +user: GithubNode.UserlikeAddress,
  +reactable: GithubNode.ReactableAddress,
|};
export type CorrespondsToCommitAddress = {|
  +type: typeof CORRESPONDS_TO_COMMIT_TYPE,
  +githubCommit: GithubNode.CommitAddress,
|};

export type StructuredAddress =
  | AuthorsAddress
  | MergedAsAddress
  | HasParentAddress
  | ReferencesAddress
  | ReactsAddress
  | CorrespondsToCommitAddress;

export const createEdge: {|
  authors: (
    author: GithubNode.UserlikeAddress,
    content: GithubNode.AuthorableAddress,
    timestampMs: TimestampMs
  ) => Edge,
  correspondsToCommit: (
    githubCommit: GithubNode.CommitAddress,
    gitCommit: GitNode.CommitAddress,
    timestampMs: TimestampMs
  ) => Edge,
  hasParent: (
    child: GithubNode.ChildAddress,
    parent: GithubNode.ParentAddress,
    timestampMs: TimestampMs
  ) => Edge,
  mergedAs: (
    pull: GithubNode.PullAddress,
    commit: GithubNode.CommitAddress,
    timestampMs: TimestampMs
  ) => Edge,
  reacts: (
    reactionType: ReactionContent,
    user: GithubNode.UserlikeAddress,
    reactable: GithubNode.ReactableAddress,
    timestampMs: TimestampMs
  ) => Edge,
  references: (
    referrer: GithubNode.TextContentAddress,
    referent: GithubNode.ReferentAddress,
    timestampMs: TimestampMs
  ) => Edge,
|} = deepFreeze({
  authors: (
    author: GithubNode.UserlikeAddress,
    content: GithubNode.AuthorableAddress,
    timestampMs: TimestampMs
  ): Edge => ({
    address: toRaw({type: AUTHORS_TYPE, author, content}),
    src: GithubNode.toRaw(author),
    dst: GithubNode.toRaw(content),
    timestampMs,
  }),
  mergedAs: (
    pull: GithubNode.PullAddress,
    commit: GithubNode.CommitAddress,
    timestampMs: TimestampMs
  ): Edge => ({
    address: toRaw({type: MERGED_AS_TYPE, pull}),
    src: GithubNode.toRaw(pull),
    dst: GithubNode.toRaw(commit),
    timestampMs,
  }),
  correspondsToCommit: (
    githubCommit: GithubNode.CommitAddress,
    gitCommit: GitNode.CommitAddress,
    timestampMs: TimestampMs
  ): Edge => ({
    address: toRaw({type: CORRESPONDS_TO_COMMIT_TYPE, githubCommit}),
    src: GithubNode.toRaw(githubCommit),
    dst: GitNode.toRaw(gitCommit),
    timestampMs,
  }),
  hasParent: (
    child: GithubNode.ChildAddress,
    parent: GithubNode.ParentAddress,
    timestampMs: TimestampMs
  ): Edge => ({
    address: toRaw({type: HAS_PARENT_TYPE, child}),
    src: GithubNode.toRaw(child),
    dst: GithubNode.toRaw(parent),
    timestampMs,
  }),
  references: (
    referrer: GithubNode.TextContentAddress,
    referent: GithubNode.ReferentAddress,
    timestampMs: TimestampMs
  ): Edge => ({
    address: toRaw({type: REFERENCES_TYPE, referrer, referent}),
    src: GithubNode.toRaw(referrer),
    dst: GithubNode.toRaw(referent),
    timestampMs,
  }),
  reacts: (
    reactionType: ReactionContent,
    user: GithubNode.UserlikeAddress,
    reactable: GithubNode.ReactableAddress,
    timestampMs: TimestampMs
  ): Edge => ({
    address: toRaw({
      type: REACTS_TYPE,
      user,
      reactionType,
      reactable,
    }),
    src: GithubNode.toRaw(user),
    dst: GithubNode.toRaw(reactable),
    timestampMs,
  }),
});

function lengthEncode(x: GithubNode.RawAddress): $ReadOnlyArray<string> {
  const parts = NodeAddress.toParts(x);
  return [String(parts.length), ...parts];
}
function lengthDecode(
  x: $ReadOnlyArray<string>,
  fail: () => Error
): {|+parts: $ReadOnlyArray<string>, +rest: $ReadOnlyArray<string>|} {
  if (x.length === 0) {
    // Not length-encoded.
    throw fail();
  }
  const [lengthString, ...allParts] = x;
  const length = parseInt(lengthString, 10);
  if (isNaN(length)) {
    throw fail();
  }
  if (length > allParts.length) {
    // Not enough elements.
    throw fail();
  }
  return {parts: allParts.slice(0, length), rest: allParts.slice(length)};
}
function multiLengthDecode(x: $ReadOnlyArray<string>, fail: () => Error) {
  let remaining = x;
  const partses = [];
  while (remaining.length > 0) {
    const {parts, rest} = lengthDecode(remaining, fail);
    partses.push(parts);
    remaining = rest;
  }
  return partses;
}

export function fromRaw(x: RawAddress): StructuredAddress {
  function fail() {
    return new Error(`Bad address: ${EdgeAddress.toString(x)}`);
  }
  if (!EdgeAddress.hasPrefix(x, GITHUB_PREFIX)) {
    throw fail();
  }
  const [_unused_sc, _unused_gh, kind, ...rest] = EdgeAddress.toParts(x);
  switch (kind) {
    case AUTHORS_TYPE: {
      const parts = multiLengthDecode(rest, fail);
      if (parts.length !== 2) {
        throw fail();
      }
      const [authorParts, contentParts] = parts;
      const author: GithubNode.UserlikeAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(authorParts): any)
      ): any);
      const content: GithubNode.AuthorableAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(contentParts): any)
      ): any);
      return ({type: AUTHORS_TYPE, author, content}: AuthorsAddress);
    }
    case MERGED_AS_TYPE: {
      const parts = multiLengthDecode(rest, fail);
      if (parts.length !== 1) {
        throw fail();
      }
      const [pullParts] = parts;
      const pull: GithubNode.PullAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(pullParts): any)
      ): any);
      return ({type: MERGED_AS_TYPE, pull}: MergedAsAddress);
    }
    case HAS_PARENT_TYPE: {
      const parts = multiLengthDecode(rest, fail);
      if (parts.length !== 1) {
        throw fail();
      }
      const [childParts] = parts;
      const child: GithubNode.ChildAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(childParts): any)
      ): any);
      return ({type: HAS_PARENT_TYPE, child}: HasParentAddress);
    }
    case REFERENCES_TYPE: {
      const parts = multiLengthDecode(rest, fail);
      if (parts.length !== 2) {
        throw fail();
      }
      const [referrerParts, referentParts] = parts;
      const referrer: GithubNode.TextContentAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(referrerParts): any)
      ): any);
      const referent: GithubNode.ReferentAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(referentParts): any)
      ): any);
      return ({type: REFERENCES_TYPE, referrer, referent}: ReferencesAddress);
    }
    case REACTS_TYPE: {
      const [rawReactionType, ...rest2] = rest;
      const reactionType = Reactions[rawReactionType];
      if (reactionType == null) {
        throw fail();
      }
      const parts = multiLengthDecode(rest2, fail);
      if (parts.length !== 2) {
        throw fail();
      }
      const [userParts, reactableParts] = parts;
      const user: GithubNode.UserlikeAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(userParts): any)
      ): any);
      const reactable: GithubNode.ReactableAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(reactableParts): any)
      ): any);
      return {type: REACTS_TYPE, reactionType, user, reactable};
    }
    default:
      throw fail();
  }
}

export function toRaw(x: StructuredAddress): RawAddress {
  switch (x.type) {
    case AUTHORS_TYPE:
      return EdgeAddress.append(
        Prefix.authors,
        ...lengthEncode(GithubNode.toRaw(x.author)),
        ...lengthEncode(GithubNode.toRaw(x.content))
      );
    case MERGED_AS_TYPE:
      return EdgeAddress.append(
        Prefix.mergedAs,
        ...lengthEncode(GithubNode.toRaw(x.pull))
      );
    case HAS_PARENT_TYPE:
      return EdgeAddress.append(
        Prefix.hasParent,
        ...lengthEncode(GithubNode.toRaw(x.child))
      );
    case REFERENCES_TYPE:
      return EdgeAddress.append(
        Prefix.references,
        ...lengthEncode(GithubNode.toRaw(x.referrer)),
        ...lengthEncode(GithubNode.toRaw(x.referent))
      );
    case REACTS_TYPE:
      return EdgeAddress.append(
        Prefix.reacts,
        x.reactionType,
        ...lengthEncode(GithubNode.toRaw(x.user)),
        ...lengthEncode(GithubNode.toRaw(x.reactable))
      );
    case CORRESPONDS_TO_COMMIT_TYPE:
      return EdgeAddress.append(
        Prefix.correspondsToCommit,
        ...lengthEncode(GithubNode.toRaw(x.githubCommit))
      );
    default:
      throw new Error((x.type: empty));
  }
}
