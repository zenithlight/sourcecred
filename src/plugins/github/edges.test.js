// @flow

import {type EdgeAddressT, edgeToParts} from "../../core/graph";
import {createEdge, fromRaw, toRaw} from "./edges";
import * as GE from "./edges";
import * as GN from "./nodes";
import {COMMIT_TYPE} from "../git/nodes";
import {ReactionContent$Values as Reactions} from "./graphqlTypes";

describe("plugins/github/edges", () => {
  const nodeExamples = {
    repo: () => ({
      type: GN.REPO_TYPE,
      owner: "sourcecred",
      name: "example-github",
    }),
    issue: () => ({
      type: GN.ISSUE_TYPE,
      repo: nodeExamples.repo(),
      number: "2",
    }),
    pull: () => ({type: GN.PULL_TYPE, repo: nodeExamples.repo(), number: "5"}),
    review: () => ({
      type: GN.REVIEW_TYPE,
      pull: nodeExamples.pull(),
      id: "100313899",
    }),
    issueComment: () => ({
      type: GN.COMMENT_TYPE,
      parent: nodeExamples.issue(),
      id: "373768703",
    }),
    pullComment: () => ({
      type: GN.COMMENT_TYPE,
      parent: nodeExamples.pull(),
      id: "396430464",
    }),
    reviewComment: () => ({
      type: GN.COMMENT_TYPE,
      parent: nodeExamples.review(),
      id: "171460198",
    }),
    user: () => ({
      type: GN.USERLIKE_TYPE,
      subtype: "USER",
      login: "decentralion",
    }),
  };

  const edgeExamples = {
    authors: () =>
      createEdge.authors(nodeExamples.user(), nodeExamples.issue(), 102),
    mergedAs: () => {
      const commit = {type: COMMIT_TYPE, id: "123"};
      return createEdge.mergedAs(nodeExamples.pull(), commit, 102);
    },
    hasParent: () =>
      createEdge.hasParent(
        nodeExamples.reviewComment(),
        nodeExamples.review(),
        102
      ),
    references: () =>
      createEdge.references(nodeExamples.issue(), nodeExamples.pull(), 102),
    reactsHeart: () =>
      createEdge.reacts(
        Reactions.HEART,
        nodeExamples.user(),
        nodeExamples.pull(),
        102
      ),
    reactsThumbsUp: () =>
      createEdge.reacts(
        Reactions.THUMBS_UP,
        nodeExamples.user(),
        nodeExamples.issue(),
        102
      ),
    reactsHooray: () =>
      createEdge.reacts(
        Reactions.HOORAY,
        nodeExamples.user(),
        nodeExamples.issueComment(),
        102
      ),
    reactsRocket: () =>
      createEdge.reacts(
        Reactions.ROCKET,
        nodeExamples.user(),
        nodeExamples.pullComment(),
        102
      ),
  };

  describe("createEdge", () => {
    Object.keys(edgeExamples).forEach((example) => {
      it(`works for ${JSON.stringify(example)}`, () => {
        const instance = edgeExamples[example]();
        expect(edgeToParts(instance)).toMatchSnapshot();
      });
    });
  });

  describe("`toRaw` after `fromRaw` is identity", () => {
    Object.keys(edgeExamples).forEach((example) => {
      it(example, () => {
        const baseAddress: EdgeAddressT = edgeExamples[example]().address;
        const instance: GE.RawAddress = (baseAddress: any);
        expect(toRaw(fromRaw(instance))).toEqual(instance);
      });
    });
  });

  describe("`fromRaw` after `toRaw` is identity", () => {
    Object.keys(edgeExamples).forEach((example) => {
      it(example, () => {
        const baseAddress: EdgeAddressT = edgeExamples[example]().address;
        const instance: GE.RawAddress = (baseAddress: any);
        const structured: GE.StructuredAddress = fromRaw(instance);
        expect(fromRaw(toRaw(structured))).toEqual(structured);
      });
    });
  });
});
