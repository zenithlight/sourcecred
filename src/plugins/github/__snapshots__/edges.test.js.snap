// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/github/edges createEdge works for "authors" 1`] = `
Object {
  "addressParts": Array [
    "sourcecred",
    "github",
    "AUTHORS",
    "5",
    "sourcecred",
    "github",
    "USERLIKE",
    "USER",
    "decentralion",
    "6",
    "sourcecred",
    "github",
    "ISSUE",
    "sourcecred",
    "example-github",
    "2",
  ],
  "dstParts": Array [
    "sourcecred",
    "github",
    "ISSUE",
    "sourcecred",
    "example-github",
    "2",
  ],
  "srcParts": Array [
    "sourcecred",
    "github",
    "USERLIKE",
    "USER",
    "decentralion",
  ],
  "timestampMs": 102,
}
`;

exports[`plugins/github/edges createEdge works for "hasParent" 1`] = `
Object {
  "addressParts": Array [
    "sourcecred",
    "github",
    "HAS_PARENT",
    "9",
    "sourcecred",
    "github",
    "COMMENT",
    "REVIEW",
    "sourcecred",
    "example-github",
    "5",
    "100313899",
    "171460198",
  ],
  "dstParts": Array [
    "sourcecred",
    "github",
    "REVIEW",
    "sourcecred",
    "example-github",
    "5",
    "100313899",
  ],
  "srcParts": Array [
    "sourcecred",
    "github",
    "COMMENT",
    "REVIEW",
    "sourcecred",
    "example-github",
    "5",
    "100313899",
    "171460198",
  ],
  "timestampMs": 102,
}
`;

exports[`plugins/github/edges createEdge works for "mergedAs" 1`] = `
Object {
  "addressParts": Array [
    "sourcecred",
    "github",
    "MERGED_AS",
    "6",
    "sourcecred",
    "github",
    "PULL",
    "sourcecred",
    "example-github",
    "5",
  ],
  "dstParts": Array [
    "sourcecred",
    "github",
    "COMMIT",
    "123",
  ],
  "srcParts": Array [
    "sourcecred",
    "github",
    "PULL",
    "sourcecred",
    "example-github",
    "5",
  ],
  "timestampMs": 102,
}
`;

exports[`plugins/github/edges createEdge works for "reactsHeart" 1`] = `
Object {
  "addressParts": Array [
    "sourcecred",
    "github",
    "REACTS",
    "HEART",
    "5",
    "sourcecred",
    "github",
    "USERLIKE",
    "USER",
    "decentralion",
    "6",
    "sourcecred",
    "github",
    "PULL",
    "sourcecred",
    "example-github",
    "5",
  ],
  "dstParts": Array [
    "sourcecred",
    "github",
    "PULL",
    "sourcecred",
    "example-github",
    "5",
  ],
  "srcParts": Array [
    "sourcecred",
    "github",
    "USERLIKE",
    "USER",
    "decentralion",
  ],
  "timestampMs": 102,
}
`;

exports[`plugins/github/edges createEdge works for "reactsHooray" 1`] = `
Object {
  "addressParts": Array [
    "sourcecred",
    "github",
    "REACTS",
    "HOORAY",
    "5",
    "sourcecred",
    "github",
    "USERLIKE",
    "USER",
    "decentralion",
    "8",
    "sourcecred",
    "github",
    "COMMENT",
    "ISSUE",
    "sourcecred",
    "example-github",
    "2",
    "373768703",
  ],
  "dstParts": Array [
    "sourcecred",
    "github",
    "COMMENT",
    "ISSUE",
    "sourcecred",
    "example-github",
    "2",
    "373768703",
  ],
  "srcParts": Array [
    "sourcecred",
    "github",
    "USERLIKE",
    "USER",
    "decentralion",
  ],
  "timestampMs": 102,
}
`;

exports[`plugins/github/edges createEdge works for "reactsRocket" 1`] = `
Object {
  "addressParts": Array [
    "sourcecred",
    "github",
    "REACTS",
    "ROCKET",
    "5",
    "sourcecred",
    "github",
    "USERLIKE",
    "USER",
    "decentralion",
    "8",
    "sourcecred",
    "github",
    "COMMENT",
    "PULL",
    "sourcecred",
    "example-github",
    "5",
    "396430464",
  ],
  "dstParts": Array [
    "sourcecred",
    "github",
    "COMMENT",
    "PULL",
    "sourcecred",
    "example-github",
    "5",
    "396430464",
  ],
  "srcParts": Array [
    "sourcecred",
    "github",
    "USERLIKE",
    "USER",
    "decentralion",
  ],
  "timestampMs": 102,
}
`;

exports[`plugins/github/edges createEdge works for "reactsThumbsUp" 1`] = `
Object {
  "addressParts": Array [
    "sourcecred",
    "github",
    "REACTS",
    "THUMBS_UP",
    "5",
    "sourcecred",
    "github",
    "USERLIKE",
    "USER",
    "decentralion",
    "6",
    "sourcecred",
    "github",
    "ISSUE",
    "sourcecred",
    "example-github",
    "2",
  ],
  "dstParts": Array [
    "sourcecred",
    "github",
    "ISSUE",
    "sourcecred",
    "example-github",
    "2",
  ],
  "srcParts": Array [
    "sourcecred",
    "github",
    "USERLIKE",
    "USER",
    "decentralion",
  ],
  "timestampMs": 102,
}
`;

exports[`plugins/github/edges createEdge works for "references" 1`] = `
Object {
  "addressParts": Array [
    "sourcecred",
    "github",
    "REFERENCES",
    "6",
    "sourcecred",
    "github",
    "ISSUE",
    "sourcecred",
    "example-github",
    "2",
    "6",
    "sourcecred",
    "github",
    "PULL",
    "sourcecred",
    "example-github",
    "5",
  ],
  "dstParts": Array [
    "sourcecred",
    "github",
    "PULL",
    "sourcecred",
    "example-github",
    "5",
  ],
  "srcParts": Array [
    "sourcecred",
    "github",
    "ISSUE",
    "sourcecred",
    "example-github",
    "2",
  ],
  "timestampMs": 102,
}
`;
